llvm/utils/TableGen/DynamicTableEmitter.cpp

```diff
@@ -0,0 +1,272 @@
+          } else
+            res << val->getAsString();
+          }
+          return res.str();
+      }
+      PrintFatalError(Loc,
+                      Twine("Entry for field '") + Field.Name + "' is null");
+    }
+    PrintFatalError(Loc, Twine("invalid field type for field '") + Field.Name + 
+                             "'; expected: bit, bits, string, or code");
+  }
+
+  void emitDynamicTable(const DynamicTable &Table, raw_ostream &OS);
+  void emitIfdef(StringRef Guard, raw_ostream &OS);
+
+  bool parseFieldType(GenericField &Field, Init *II);
+  void collectTableEntries(DynamicTable &Table,
+                           const std::vector<Record *> &Items);
+};
+
+} // End anonymous namespace.
+
+void DynamicTableEmitter::emitIfdef(StringRef Guard, raw_ostream &OS) {
+  OS << "#ifdef " << Guard << "\n";
+  PreprocessorGuards.insert(std::string(Guard));
+}
+
+void DynamicTableEmitter::emitDynamicTable(const DynamicTable &Table,
+                                           raw_ostream &OS) {
+  emitIfdef((Twine("GET_") + Table.PreprocessorGuard + "_IMPL").str(), OS);
+
+  // The primary data table contains all the fields defined for this map.
+  OS << "std::map<std::string, " << Table.CppTypeName << "> " << Table.Name
+     << " = {\n";
+  for (unsigned i = 0; i < Table.Entries.size(); ++i) {
+    Record *Entry = Table.Entries[i];
+    OS << "  { ";
+
+    ListSeparator LS;
+    bool first = true;
+    bool second = false;
+    for (const auto &Field : Table.Fields) {
+      std::string val = primaryRepresentation(Table.Locs[0], Field,
+                                              Entry->getValueInit(Field.Name));
+      if (first) {
+        first = false;
+        second = true;
+        OS << LS << val << ", { ";
+      } else {
+        if (!second) {
+          OS << LS;
+        }
+        if (Field.IsList) {
+          val.insert(0, 1, '{');
+          val.push_back('}');
+        }
+        OS << val;
+        second = false;
+      }
+    }
+
+    OS << " }}, // " << i << "\n";
+  }
+  OS << " };\n";
+
+  OS << "#endif\n\n";
+}
+
+bool DynamicTableEmitter::parseFieldType(GenericField &Field, Init *TypeOf) {
+  if (auto Type = dyn_cast<StringInit>(TypeOf)) {
+    if (Type->getValue() == "code") {
+      Field.IsCode = true;
+      return true;
+    } else if (Type->getValue().starts_with("list")) {
+      // Nested lists are not allowed, make sure there are none
+      auto occurrences = Type->getValue().count("list");
+      if (occurrences > 1) {
+        PrintFatalError(Twine("Nested lists are not allowed: ") +
+                        Type->getValue().str());
+      }
+      Field.IsList = true;
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void DynamicTableEmitter::collectTableEntries(
+    DynamicTable &Table, const std::vector<Record *> &Items) {
+  if (Items.empty())
+    PrintFatalError(Table.Locs,
+                    Twine("Table '") + Table.Name + "' has no entries");
```

Comment:
I don't have a use case for an empty table, but I'm not sure if that should necessary be a fatal error

Reply:
This comes from the searchable tables backend. I guess for that backend, there's no point in having empty tables, since they can't be extended dynamically. In our backend, given that this is a map, and so, can be extended dynamically, one could argue that empty tables could have a use. I will allow this for dynamic tables.